kale/task.py

```diff
@@ -295,3 +355,57 @@ def _report_permanent_failure(cls, message, exception, failure_type,
                        (failure_type, message.task_name, message.task_id,
                         exception))
         logger.error(message_str)
+
+
+class RemoteException(object):
+    """Utility class for reraising an exception from a remote process.
+
+    Keeps the original stack trace.
+    """
+    def __init__(self):
+        self.exc_info = sys.exc_info()
+
+    def reraise(self):
+        six.reraise(*self.exc_info)
+
+
+def _task_wrapper(task_inst, args, kwargs):
+    """Task wrapper function.
+
+    In single-process mode, this will return any exception raised.
+
+    In multi-process mode, this will return a RemoteException for any exceptions raised.
+    """
+    # This enables tasks to be blacklisted by ID or type.
+    task_inst._setup_task_environment()
+    task_inst._pre_run(*args, **kwargs)
+
+    if settings.is_multiprocess_mode():
```

Comment:
Create a separate wrapper for multi/single or no wrapper at all for single. Checking this setting in multiple places (here multiple times and in the caller) makes this more complex and prone to error.